pages/controls/Certificate_and_Public_Key_Pinning.md

```diff
@@ -357,40 +357,161 @@ However, implementing pinning validation from scratch should be avoided,
+
+In .NET Framework 4.5 onwards, a validation callback can be set per HTTP request
+
+```
+// Domain name for which to perform certificate pinning
+private const string DomainName = "encrypted.google.com";
 
-  String pk = certificate.GetPublicKeyString();
-  if (pk.Equals(PUB_KEY))
-    return true;
+// Encoded RSAPublicKey for Domain name
+private const string DomainPublicKey = "30818902818100C4A06B7B52F8D17DC1CCB47362" +
+	"C64AB799AAE19E245A7559E9CEEC7D8AA4DF07CB0B21FDFD763C63A313A668FE9D764E" +
+	"D913C51A676788DB62AF624F422C2F112C1316922AA5D37823CD9F43D1FC54513D14B2" +
+	"9E36991F08A042C42EAAEEE5FE8E2CB10167174A359CEBF6FACC2C9CA933AD403137EE" +
+	"2C3F4CBED9460129C72B0203010001";
+
+public static void Main(string[] args)
+{
+	var wr = (HttpWebRequest)WebRequest.Create(string.Format("https://{0}/", DomainName));
+	
+	// set server validation callback for this request only
+	wr.ServerCertificateValidationCallback = PinPublicKey;
+	wr.GetResponse();
+}
+
+private static bool PinPublicKey(
+	object sender,
+	X509Certificate certificate,
+	X509Chain chain,
+	SslPolicyErrors sslPolicyErrors)
+{
+	if (certificate == null)
+		return false;
+
+	var request = sender as HttpWebRequest;
+	if (request == null) 
+		return false;
+		
+	// if the request is for the target domain, perform certificate pinning
+	if (string.Equals(request.Address.Authority, DomainName, StringComparison.OrdinalIgnoreCase))
+	{
+		var pk = certificate.GetPublicKeyString();
+		return pk.Equals(DomainPublicKey);
+	}
+
+	// Check whether there were any policy errors for any other domain
+	return sslPolicyErrors == SslPolicyErrors.None;
+}
+```
+
+#### HttpClient
+
+When using `HttpClient`, a server validation callback can be set on
+`HttpClientHandler` that will affect **all** requests requiring
+validation made with `HttpClient` instances using the `HttpClientHandler`.
+
+```
+// Domain name for which to perform certificate pinning
+private const string DomainName = "encrypted.google.com";
+
+// Encoded RSAPublicKey for Domain name
+private const string DomainPublicKey = "30818902818100C4A06B7B52F8D17DC1CCB47362" +
+	"C64AB799AAE19E245A7559E9CEEC7D8AA4DF07CB0B21FDFD763C63A313A668FE9D764E" +
+	"D913C51A676788DB62AF624F422C2F112C1316922AA5D37823CD9F43D1FC54513D14B2" +
+	"9E36991F08A042C42EAAEEE5FE8E2CB10167174A359CEBF6FACC2C9CA933AD403137EE" +
+	"2C3F4CBED9460129C72B0203010001";
+
+public static void Main(string[] args)
+{
+	MainAsync().Wait();
+}
+
+private static async Task MainAsync()
+{
+	var handler = new HttpClientHandler
+	{
+		ServerCertificateCustomValidationCallback = PinPublicKey
+	};
+
+	var client = new HttpClient(handler);
+	await client.GetAsync(string.Format("https://{0}/", DomainName));
+}
 
-  // Bad dog
-  return false;
+private static bool PinPublicKey(
+	HttpRequestMessage requestMessage,
+	X509Certificate2 certificate,
+	X509Chain chain,
+	SslPolicyErrors sslPolicyErrors)
+{~~~~
```

Comment:
@russcam there's some wayward tildes here

Reply:
not sure how those got there 😄 Will update now

Reply:
Done